File: editor/src/main/com/mbrlabs/mundus/ui/modules/dialogs/importer/ImportTextureDialog.java
/*
 * Copyright (c) 2016. See AUTHORS file.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mbrlabs.mundus.ui.modules.dialogs.importer;

import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.Disposable;
import com.kotcrab.vis.ui.widget.VisLabel;
import com.kotcrab.vis.ui.widget.VisTable;
import com.kotcrab.vis.ui.widget.VisTextButton;
import com.kotcrab.vis.ui.widget.VisTextField;
import com.mbrlabs.mundus.commons.assets.TextureAsset;
import com.mbrlabs.mundus.assets.EditorAssetManager;
import com.mbrlabs.mundus.core.Inject;
import com.mbrlabs.mundus.core.Mundus;
import com.mbrlabs.mundus.core.project.ProjectManager;
import com.mbrlabs.mundus.core.registry.Registry;
import com.mbrlabs.mundus.ui.Ui;
import com.mbrlabs.mundus.ui.modules.dialogs.BaseDialog;
import com.mbrlabs.mundus.ui.widgets.ImageChooserField;
import com.mbrlabs.mundus.utils.FileFormatUtils;

/**
 * @author Marcus Brummer
 * @version 07-06-2016
 */
public class ImportTextureDialog extends BaseDialog implements Disposable {

    private ImportTextureTable importTextureTable;

    @Inject
    private Registry registry;
    @Inject
    private ProjectManager projectManager;

    public ImportTextureDialog() {
        super("Import Texture");
        Mundus.inject(this);
        setModal(true);
        setMovable(true);

        Table root = new VisTable();
        add(root).expand().fill();
        importTextureTable = new ImportTextureTable();

        root.add(importTextureTable).minWidth(600).expand().fill().left().top();
    }

    @Override
    public void dispose() {
        importTextureTable.dispose();
    }

    /**
     *
     */
    private class ImportTextureTable extends VisTable implements Disposable {
        // UI elements
        private VisTextField name = new VisTextField();
        private VisTextButton importBtn = new VisTextButton("IMPORT");
        private ImageChooserField imageChooserField = new ImageChooserField(200);


        public ImportTextureTable() {
            super();
            this.setupUI();
            this.setupListener();

            align(Align.topLeft);
        }

        private void setupUI() {
            padTop(6).padRight(6).padBottom(22);

            VisTable left = new VisTable();
            VisTable right = new VisTable();

            left.add(new VisLabel("Name: ")).left().row();
            left.add(name).expandX().fillX().row();
            left.add(importBtn).fillX().expand().bottom();

            right.add(imageChooserField);

            add(left).width(300).top().left().expandY().fillY();
            add(right).width(300);

        }

        private void setupListener() {
            importBtn.addListener(new ClickListener() {
                @Override
                public void clicked(InputEvent event, float x, float y) {
                    FileHandle texture = imageChooserField.getFile();
                    if(texture != null && texture.exists() && FileFormatUtils.isImage(texture)) {
                        EditorAssetManager assetManager = projectManager.current().assetManager;
                        assetManager.importAsset(texture, TextureAsset.class);
                        close();
                        Ui.getInstance().getToaster().success("Texture imported");
                    } else {
                        Ui.getInstance().getToaster().error("There is nothing to import");
                    }
                }
            });
        }


        @Override
        public void dispose() {

        }
    }

}

File: editor/src/main/com/mbrlabs/mundus/editor/ui/modules/dialogs/importer/ImportTextureDialog.kt
/*
 * Copyright (c) 2016. See AUTHORS file.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mbrlabs.mundus.editor.ui.modules.dialogs.importer

import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.Disposable
import com.kotcrab.vis.ui.widget.VisTable
import com.kotcrab.vis.ui.widget.VisTextButton
import com.mbrlabs.mundus.editor.Mundus
import com.mbrlabs.mundus.editor.assets.AssetAlreadyExistsException
import com.mbrlabs.mundus.editor.core.project.ProjectManager
import com.mbrlabs.mundus.editor.events.AssetImportEvent
import com.mbrlabs.mundus.editor.ui.UI
import com.mbrlabs.mundus.editor.ui.modules.dialogs.BaseDialog
import com.mbrlabs.mundus.editor.ui.widgets.ImageChooserField
import com.mbrlabs.mundus.editor.utils.Log
import com.mbrlabs.mundus.editor.utils.isImage
import java.io.IOException

/**
 * @author Marcus Brummer
 * @version 07-06-2016
 */
class ImportTextureDialog : BaseDialog("Import Texture"), Disposable {

    companion object {
        private val TAG = ImportTextureDialog::class.java.simpleName
    }

    private val importTextureTable: ImportTextureTable

    private val projectManager: ProjectManager = Mundus.inject()

    init {
        isModal = true
        isMovable = true

        val root = VisTable()
        add<Table>(root).expand().fill()
        importTextureTable = ImportTextureTable()

        root.add(importTextureTable).minWidth(300f).expand().fill().left().top()
    }

    override fun dispose() {
        importTextureTable.dispose()
    }

    override fun close() {
        super.close()
        importTextureTable.removeTexture()
    }

    /**

     */
    private inner class ImportTextureTable : VisTable(), Disposable {
        // UI elements
        private val importBtn = VisTextButton("IMPORT")
        private val imageChooserField = ImageChooserField(300)

        init {
            this.setupUI()
            this.setupListener()

            align(Align.topLeft)
        }

        private fun setupUI() {
            padTop(6f).padRight(6f).padBottom(22f)
            add(imageChooserField).grow().row()
            add(importBtn).grow().row()
        }

        private fun setupListener() {
            importBtn.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent?, x: Float, y: Float) {
                    try {
                        val texture = imageChooserField.file
                        if (texture != null && texture.exists() && isImage(texture)) {
                            val assetManager = projectManager.current().assetManager
                            val asset = assetManager.createTextureAsset(texture)
                            Mundus.postEvent(AssetImportEvent(asset))
                            close()
                            UI.toaster.success("Texture imported")
                        } else {
                            UI.toaster.error("There is nothing to import")
                        }
                    } catch (e: IOException) {
                        Log.exception(TAG, e)
                        UI.toaster.error("IO error")
                    } catch (ee: AssetAlreadyExistsException) {
                        Log.exception(TAG, ee)
                        UI.toaster.error("Error: There already exists a texture with the same name")
                    }

                }
            })
        }

        fun removeTexture() {
            imageChooserField.removeImage()
        }

        override fun dispose() {

        }
    }

}
